core/net/rpl/rpl-ext-header.c

```diff
@@ -206,8 +208,10 @@ rpl_srh_get_next_hop(uip_ipaddr_t *ipaddr)
     }
   }
 
-  if(uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
-      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) {
+  if((uip_next_hdr != NULL && *uip_next_hdr == UIP_PROTO_ROUTING
+      && UIP_RH_BUF->routing_type == RPL_RH_TYPE_SRH) ||
+     (default_instance->current_dag->rank == ROOT_RANK(default_instance) &&
+      rpl_ns_get_node(dag, &UIP_IP_BUF->destipaddr) != NULL)) {
```

Comment:
There are two cases where `rpl_srh_get_next_hop()` should convert the global address in a destination address field to a correspondent link-local address:
 - Case-1. Any node in a RPL network processes a packet which has a source routing header.
 - Case-2. The root node processes a packet (originally) destined for one of its children.

The first case is handled by the original code. The second one is what I tried to handle by this change. The point is that such a packet as described in Case-2 does does not have a source routing header. And 'rpl_srh_get_next_hop()' does nothing for a packet without the header.

The additional condition that is combined with the original one by `||` consists of two parts:

- 2-1. `default_instance->current_dag->rank == ROOT_RANK(default_instance)`
- 2-2. `rpl_ns_get_node(dag, &UIP_IP_BUF->destipaddr) != NULL`

2-1 determines if the node is the root node. 2-2 determines if the destination is in the RPL network. If a more strict check is required, we need to check if the return value from `rpl_ns_get_node()` has `root_node` as its parent.